src/ray/raylet/node_manager.cc

```diff
@@ -1633,11 +1673,40 @@ void NodeManager::HandleReturnWorker(const rpc::ReturnWorkerRequest &request,
   std::shared_ptr<Worker> worker = std::move(leased_workers_[worker_id]);
 
   if (new_scheduler_enabled_) {
+    if (worker->IsBlocked()) {
+      // If worker blocked, unblock it to return the cpu resources back to the worker.
+      HandleDirectCallTaskUnblocked(worker);
+    }
     auto it = leased_worker_resources_.find(worker_id);
     RAY_CHECK(it != leased_worker_resources_.end());
-    new_resource_scheduler_->AddNodeAvailableResources(client_id_.Binary(), it->second);
+
+    new_resource_scheduler_->AddNodeAvailableResources(client_id_.Binary(),
+      it->second.GetResourceMap());
+
+    if (worker->borrowed_cpu_resources_.GetResourceMap().size()) {
+      // This machine is oversubscribed, so the worker didn't get back cpus when unblocked.
+      // Thus we need to substract these cpus, as the previous "AddNodeAvailableResources" call
+      // assumed they were allocated to this worker.
+      new_resource_scheduler_->SubtractNodeAvailableResources(client_id_.Binary(),
+          worker->borrowed_cpu_resources_.GetResourceMap());
+      worker->borrowed_cpu_resources_ = ResourceSet();
+    }
```

Comment:
I found this concept of a borrowed resource confusing, is there any way to avoid it?

Reply:
I can see that. I put a note so we can improve this in the next PR.